Viewing party project completion:

MBsea21 opened the pull request:
Apologies on the delay. I was unable to really allocate the right time/ brain power to focus on part 3+ of the project till today. I am looking forward to doing some more pair programming in the future to continue building up those skill sets and am bummed I couldn't work with a partner for this project but completely understand the reason why! Once I had a full nights rest I was able to wrap my head around the complexities of the project. 

I made some adjustments to my code once I have all tests pass. One was creating a helper function that compiles the movies in friends lists into one list to help streamline future functions. It helped prevent risk of typing in a bug by making the movie compilation a function call instead of having to write out the iterations each time. 

now that I have completed I will be sure to watch the viewing party recap and will update absent @ ada with my recap of the things I learned/would do differently.

MBsea21 committed:
Wave one done, all tests pass

MBsea21 committed:
Wave 2 has all tests passed

MBsea21 committed:
wave 3 all tests passed and updated as needed

MBsea21 committed:
wave 4 all tests passed

MBsea21 committed:
wave 5 all tests passed

MBsea21 committed:
refactored wave 3 to include join_friends_movie_list helper function

MBsea21 committed:
general spacing updates. working on making code more readable.

anselrognlie commented on tests/test_wave_01.py:
```diff
@@ -157,14 +159,16 @@ def test_moves_movie_from_watchlist_to_empty_watched():
+    assert len(updated_data["watched"]) == 1   
+    assert updated_data["watched"][0]["title"] == MOVIE_TITLE_1
+    assert updated_data["watched"][0]["genre"] == GENRE_1
+    assert updated_data["watched"][0]["rating"] == RATING_1
```
👍 We should check all three parts of the movie dict.

Here, we are certain that the movie is in position 0 of the watched list (since there's only a single movie). But since the project doesn't specify where in the watched list the movie should appear, we might consider performing an `in` check, with a movie dict that we construct with the expected values, in cases where other data might be present.

MBsea21 replied:
Updated this in the code and ran tests to ensure everything passes ok.

anselrognlie commented on tests/test_wave_01.py:
```diff
@@ -182,13 +186,13 @@ def test_moves_movie_from_watchlist_to_watched():
-    # *******************************************************************************************
+    assert updated_data["watched"][1]["title"] == MOVIE_TITLE_1
+    assert updated_data["watched"][1]["genre"] == GENRE_1
+    assert updated_data["watched"][1]["rating"] == RATING_1 
```
👍 We should check all three parts again, and while your implementation _does_ append to the watched list, since we don't have that specified, we might stick with an `in` check so that doesn't become an assumption in the test (which could restrict our options to refactor later). Here, we also have a variable the refers to the movie that should have moved, which we can reuse.

```py
assert movie_to_watch in updated_data["watched"]
```

Additionally, it would be a good idea to check that the movie remaining in the watchlist and the movie that was already watched are still in the lists we expect to find them in.

anselrognlie commented on tests/test_wave_03.py:
```diff
@@ -54,13 +54,13 @@ def test_friends_unique_movies_not_duplicated():
 
     # Assert
     assert len(friends_unique_movies) == 3
+    assert INTRIGUE_3 in friends_unique_movies
+    assert HORROR_1 in friends_unique_movies
+    assert FANTASY_4 in friends_unique_movies
```
👍 Great way to check that the three movies are in the output.

Here, it's even more important not to assume a particular ordering of the outputs. The act of making some data unique can often result in data ending up in arbitrary locations relative to where they occurred in the original data, especially if intermediate data structures like sets or dicts are involved.

anselrognlie commented on tests/test_wave_05.py:
```diff
@@ -53,12 +53,13 @@ def test_new_genre_rec_from_empty_friends():
 
-@pytest.mark.skip()
+#Assert
+    assert len(reccomendations) == 0
```
👍 This test needed both the act and the assert.

anselrognlie replied:
Nit: align the comments to the code and include a space after the `#`
```py
    # Act
    reccomendations = get_new_rec_by_genre(sonyas_data)

    # Assert
    assert len(reccomendations) == 0
```

anselrognlie commented on viewing_party/__init__.py:
👀 No need to include a blank line in this file. Be sure to use `git status` to make sure we're not staging/committing unintentional changes.

`git status` shows the commands to use to move a file from staged to unstaged, and unstaged to reverting the change.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
 # ------------- WAVE 1 --------------------
+# create dictionary with following key value pairs. 
+import pprint
```
👀 Remove imports that were being used for debugging output before submitting.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+    elif genre is None: 
+        return None
+    elif rating is None: 
+        return None
```
👀 The specification says to return `None` if any of these values are falsy, which is a wider range than checking for `None`.
```py
    if not title:  # checks for falsy title 
        return None
```

Since these checks are pretty short, consider grouping them together into a single condition using `or`.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+        return None
+    elif rating is None: 
+        return None
+    movie_dictionary = {"title": title, "genre": genre, "rating": rating }
```
Consider making CONSTANTs to hold the key strings. This can help us get code completion suggestions and also avoid potential typos.
_(earlier in the file)_
```py
TITLE = "title"
GENRE = "genre"
RATING = "rating"
```
_(later)_
```py
    movie_dictionary = {
        TITLE: title,
        GENRE: genre,
        RATING: rating,
    }
```

This could be done for all of the keys we'll be using.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+
+
+def add_to_watched(user_data,movie):
+    watched_list = user_data["watched"]
```
👍 If I'm going to be accessing a nested data value more than a few times, I do like to make a local named reference to it for simplicity (and to avoid the possibility of errors each time I rewrite the nested expression.

Though here, I might access the value directly, since we're only appending to it once.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+    watch_list = user_data["watchlist"]
+    watch_list.append(movie)
+    return user_data
     pass
```
👀  Remove the placeholder `pass` statements

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
 
+def watch_movie(user_data, movie):
+    watch_list = user_data["watchlist"]
+    print("watch list :", user_data["watchlist"])
+    print("watched list :", user_data["watched"])
```
👀 Remove debug print statements before submitting.

Also, to get nicely "labelled" print output, we can use a minor format string variant
```py
    print(f'{user_data["watchlist"]=}')  # note the trailing `=`
```
This will literally print the text of the expression, followed by the value of the expression.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+    return user_data
     pass
 
+def watch_movie(user_data, movie):
```
👀 Prefer a name like `title` or `movie_title` rather than movie. The input here is a title string, not a full movie dict, so a name that reminds us of that would be helpful.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+    for to_watch_movie in watch_list: 
+        movie_name = to_watch_movie["title"]
+        if movie_name == movie:
+            watch_list.remove(to_watch_movie)
```
We should be very careful about modifying a collection over which we are currently iterating. It ends up not mattering here, since our examples expect only a single value to be removed, so any odd iteration behavior after that doesn't cause a problem (note that some languages will intentionally crash when they detect that a modification has occurred to a collection, and an iteration over the collection continues).

To make it very clear that we are following safe behaviors, we should either:
1. Terminate the loop as soon as we have completed the modification, or
2. Only use the loop to _find_ the value we want to remove (either the entire value, if we later use `remove`, or the position, if we later use `pop`, and then remove the value after the loop completes.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
```
👍 NIce guard clause, summing logic, and average calculation. I'd just like a few blank lines to separate the main chunks of the logic.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+            top_genre = genre_name
+            top_genre_count = genre_count
+        elif genre_count > top_genre_count: 
+            top_genre = genre_name
```
👀 As written, there's a bug here that could allow later values in the map to overwrite the correct value, since the count won't be bumped up to reflect the max genre (unless that genre is also the first in the map). Essentially, the count will be set once, and the last genre with a higher (not tied) count will end up being picked as the max.

If we address this, we might also be able to simplify the conditions here a bit.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+        elif genre_count > top_genre_count: 
+            top_genre = genre_name
+        else: 
+            continue
```
👀 Since there's nothing after this in the loop, an `else` with `continue` has no effect and should be omitted. If there were additional code in the loop after the conditions that we wanted to skip, then this might be appropriate, but I would usually try to reogranize my checks to avoid something like this.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+    for movie in watched_movies:
+        genre = movie["genre"]
+        if genre not in genre_dict.keys():
+            genre_dict.update({genre:1})
```
👀 To add a new key into the dictionary, all we need to do is assign to it. `update` is more for the case where we have two existing dictionaries and want to merge the keys from one into the other. Here
```py
            genre_dict[genre] = 1
```
will get the job done.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+        sum_ratings += movie_rating
+    return sum_ratings / count_movies_watched
 
+def get_most_watched_genre(user_data):
```
👍 Nice approach of using a guard clause, building a frequency map, then finding the max count.

anselrognlie replied:
👀 There is a small (but significant) bug in the max calculation. See my note below.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+                unique_watched.append(friends_movie)
+    return unique_watched
+
+def join_friends_list(user_data):
```
👍 Nice helper to produce a flattend list of movies watched by the friends.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+    return unique_watched
+
+def join_friends_list(user_data):
+    friends_list = user_data.get("friends")
```
`get` is mostly used when there's a reasonable default that we can use even if the specified key is missing. If we don't specify the default, python uses `None`. This isn't a huge help to us here, since line 109 would crash when it tries to iterate over `None`.

I'd probably just use `[]`-based access of the `"friends"` key. True, it would crash if there were no friends data, but that also probably means something has gone terribly wrong with our system.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+    friends_movie_list = join_friends_list(user_data)
+
+    for my_movie in my_movies:
+        if my_movie not in friends_movie_list and my_movie not in unique_watched:
```
Nice checks to ensure that we filter out movies the friends have also watched, and only add the user movies to the result a single time. It might be safe to assume there will be no duplicates in the user watched list (though not necessarily clear from the README; is it a watch _history_ list, or an _ever_ watched list), but doing the check doesn't hurt.

However, keep in mind that each of those `not in` checks must iterate through the list it's checking against. This might be a long list for the firends list (though this isn't _exactly_ a O(n^2) situation, that's what it resembles), and while the result list starts small, as we add to it it gets longer, with each new movie needing to check against each previous movie. This _is_ O(n^2).

Instead of checking directly checking against lists of movies, we might use one of our data structures with constant lookup (dict/set) and choose some immutable piece of data from each movie to stand in for the entire record. From the earlier implementation of `watch_movie`, we appear to treat a movie's title as uniquely representing that movie, so we could build sets to track the titles of all the friend movies, and the movies we've added to the output list so far. Checking a new candidate movie against these sets is constant time, avoiding the trend towards O(n^2) performance.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+def get_unique_watched(user_data):
+    unique_watched = []
+    my_movies = user_data.get("watched")
+    friends_movie_list = join_friends_list(user_data)
```
👍 Nice use of your helper to get the merged list of all movies watched by the friends.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+    return unique_watched
+    
+
+def get_friends_unique_watched(user_data):
```
The feedback for `get_unique_watched` applies here as well.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+    movie_recs = []
+    unique_movies=get_friends_unique_watched(user_data)
+    for movie in unique_movies:
+        movie_host = movie.get("host")
```
The use of `get` here feels a bit more natural, as we've definitely seen that `"host"` wasn't present in earlier waves. That being said, I think that's more a product of not wanting to have the earlier waves feel overwhelming with _all_ the data present right from the start. As a result, I would still tend to stick with `[]` notation to retrieve the host value, as I like when unexpected data results in a crash, without clear direction that it should be considered optional.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
 
+def get_available_recs(user_data):
+    movie_recs = []
+    unique_movies=get_friends_unique_watched(user_data)
```
👍 Nice use of the previous wave function so that we can start with a list of movies only the friends have watched. All we need to do then is find those whose host appears in the user subscription list.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+    unique_movies=get_friends_unique_watched(user_data)
+    for movie in unique_movies:
+        movie_host = movie.get("host")
+        if movie_host in user_data["subscriptions"]:
```
Although we'd expect the subscription list to be relatively short, it still might be worth loading the data into a set, since that would get us constant time `in` lookups.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+    joined_friends_list = join_friends_list(user_data)
+    rec_from_favorites_list = []
+
+    print(joined_friends_list)
```
Nit: remove debug print statements

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+    new_rec_by_genre = []
+    joined_friends_list = join_friends_list(user_data)
+    if user_data["watched"] == [] or joined_friends_list == []:
+        return new_rec_by_genre
```
While it may seem like a point of efficiency to check whether there are any movies in the user or friends list, and quit early, keep in mind that when there _is_ friend movie data, this results in building a merged list of the movies (which could be large) and is then discarded without further use.

I do see that without this check, one of the tests fails while trying to access the subscription data of the user, but this is because this function was not specified to take subscription/host information into account when making a genre recommendation. In actuality, the first test (the only test of `get_new_rec_by_genre` that returns a non-empty result) is passing due to a fluke in the data, that the host/subscription data in the less pared down structure _does_ have host and subscription data (it was based on the data from wave 4), and that data happens to include the expected genre recommendation result in the `get_available_recs` result. This is an error in implementing the specification.

We _do_ have a previous function which, if dropped in in place of the call to `get_available_recs` does match the specified behavior, and also passes the tests without this initial check.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+    
+
+    most_watched_genre = get_most_watched_genre(user_data)
+    unreviewed_recs = get_available_recs(user_data)
```
👀 Using `get_available_recs` here conflicts with the specified wave behavior. `get_available_recs` applies host/subscription data when looking for movies watched by the friends, but not by the user, but `get_new_rec_by_genre` is specified to consider a movie to be part of the result if:
  - The user has not watched it
  - At least one of the user's friends has watched
  - The `"genre"` of the movie is the same as the user's most frequent genre
There is no requirement about host/subscription data.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+    print(joined_friends_list)
+    for movie in favorites:
+        if movie not in joined_friends_list: 
+            rec_from_favorites_list.append(movie)
```
👍 Nice logic to filter the favorites list of movies against the list of movies watched by the friends.

Notice that this logic duplicates logic we've previously implemented, filtering against the movies watched by friends. What if we started with the list of movies watched by the user, but not by the friends (it would be reasonable to assume that a user wouldn't have a movie as a favorite if they haven't watched it)? We already have logic that performs this part of the calculation. Then we could just filter that list against movies in the favorite list.

As with other similar logic elsewhere, it would be worth considering using an auxilliary set of titles to help reduce the cost of checking whether a particular movie is in a certain list.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+
+def get_rec_from_favorites(user_data):
+    favorites = user_data["favorites"]
+    joined_friends_list = join_friends_list(user_data)
```
👍 Nice reuse of your joined friend movie helper.

But if we change our perspective a little bit, we might be able to reuse a different function that does some of the same logic also performed later in this function. See my note below.

anselrognlie commented on viewing_party/party.py:
```diff
@@ -1,23 +1,158 @@
+    top_genre = None
+    top_genre_count = 0
+    if watched_movies == []:
+        return None
```
Do we need the dedicated guard clause here? If there were no watched movies, what would your logic produce for a result?

anselrognlie reviewed (commented):
I've added comments about the implementation as submitted. Please review them, and let me know if you have any questions. There are a few things I'd like you to revisit before we move on from this project. My Slack communication has more detail about exactly what to focus on.

MBsea21 committed:
"corrected the following bugs / issues

removed join_friends helper function: helper function was adding unnecessary space complexity.

updated functions that called on joined friends function to work without helper function.

updated wave 1 assert statements to not includee list location and use in method instead.

removed erraneous space in __init___ file.

adjusted formatting in test wave 5

Made sure to remove redudency by calling previouusly created functions where neccessary. (ex: get_friends_unique()).

updated get_new_rec_by_genre result to not be dependent on subscriptions by changing the program to look at get_friends_unique_watched rather than get_available_recs

Removed erraneous guard claus that checked if list was empty

"

MBsea21 committed:
"updated get rec from favorites to use set of movie titles to reduce complexity"

MBsea21 committed:
corrected top genre bug which did not update top genre count.

MBsea21 committed:
updated code according to advice provided in pull request. all tests passed.

MBsea21 commented:
@anselrognlie I have updated the code and I think I was able to address the points you mentioned in your pull request! Apologies if I missed anything.

MBsea21 committed:
updated validation sequence at start of code to ensure the data is valid.

MBsea21 commented:
Just committed a new change to update the validation function at the beginning of the code.

MBsea21 committed:
updated test one to use in functionality rather than indexing on location

MBsea21 committed:
updated wave 1
